objects/src/accounts/storage/mod.rs

```diff
@@ -325,20 +360,27 @@ impl AccountStorage {
                         actual: value_arity,
                     });
                 }
-
-                // make sure the value matches the root of the corresponding storage map
-                // TODO: we should remove handling of storage map updates from set_item();
-                // once this is done, these checks would not be necessary
-                let storage_map = self.maps.get(&index).expect("storage map not found");
-                assert_eq!(storage_map.root(), value.into());
             },
-            slot_type => Err(AccountError::StorageSlotArrayNotAllowed(index, slot_type))?,
+            slot_type => Err(AccountError::MapsUpdateToNonMapsSlot(index, slot_type))?,
         }
 
-        // update the slot and return
+        // get the correct map
+        let storage_map =
+            self.maps.get_mut(&index).ok_or(AccountError::StorageMapNotFound(index))?;
+
+        // get old values to return
+        let old_map_root = storage_map.root();
+        let old_value = storage_map.get_value(&Digest::from(key));
+
+        // apply the delta
+        storage_map.insert(key.into(), value);
+
+        // update the root of the storage map in the corresponding storage slot
+        let new_root = storage_map.root().into();
         let index = LeafIndex::new(index.into()).expect("index is u8 - index within range");
-        let slot_value = self.slots.insert(index, value);
-        Ok(slot_value)
+        self.slots.insert(index, new_root);
+
+        Ok((old_map_root.into(), old_value))
```

Comment:
What's the reason for returning both the old root and the old value? Could we get away with returning just the old value? If so, this could look something like this:
```Rust
let old_value = storage_map.insert(key.into(), value);

// update the root of the storage map in the corresponding storage slot
let index = LeafIndex::new(index.into()).expect("index is u8 - index within range");
self.slots.insert(index, storage_map.root().into());

Ok(old_value)
```